bpo-42128: Add documentation for pattern matching (PEP 634)

dmoisset opened the pull request:
This covers documentation for the new match statement in 3.10, as requested by the SC in https://mail.python.org/archives/list/python-committers@python.org/message/SQC2FTLFV5A7DV7RCEAR2I2IKJKGK7W3/

The documentation includes:
* A large language reference update include syntax and behaviour of the match statement, largely based on the relevant sections of PEP-634 in the compound statements section.
* A significant new section in the "control flow" chapter of the tutorial based on the Appendix A of PEP-636
* An update to the "data model" reference chapter describing the new special attribute `__match_args__`
* A new section in the "lexical analysis" chapter describing the use of soft keywords.
* An update to the FAQ entry about "Why isn't there a switch or case statement in Python?"

<!-- issue-number: [bpo-42128](https://bugs.python.org/issue42128) -->
https://bugs.python.org/issue42128
<!-- /issue-number -->

Fidget-Spinner committed:
Add half of PEP 634 (lots of TODOs left)

Co-Authored-By: Daniel F Moisset <dfmoisset@gmail.com>
Co-Authored-By: Guido van Rossum <2894642+gvanrossum@users.noreply.github.com>
Co-Authored-By: Brandt Bucher <brandt@python.org>

Fidget-Spinner committed:
use better wording and add lots of lists and tables (still lots of TODOs left)

one table idea was inspired by Raymond's awesome tweet here https://twitter.com/raymondh/status/1361780586570948609

Co-Authored-By: Raymond Hettinger <1623689+rhettinger@users.noreply.github.com>

dmoisset committed:
Add section on soft keywords "match" and "case"

dmoisset committed:
Add tutorial on pattern matching

The tutorial is a slightly modified version of PEP-636 Appendix A

Fidget-Spinner committed:
done!

Fidget-Spinner committed:
move index and link down to correct (original) location,  add index for soft keywords

dmoisset committed:
Add "_" as a soft keyword

dmoisset committed:
Add __match_args__ to the data model

Fidget-Spinner committed:
Apply suggested changes from code review by Daniel (part 1)

Co-authored-by: Daniel F Moisset <dfmoisset@gmail.com>

Fidget-Spinner committed:
Apply final suggestions

Fidget-Spinner committed:
Add link, make grammar and links consistent with rest of docs

Fidget-Spinner committed:
touch up table

Fidget-Spinner committed:
Add versionadded to datamodel

Fidget-Spinner committed:
Rename __match_args__ section link

Fidget-Spinner committed:
object.__match_args__ -> class.__match_args__

Fidget-Spinner committed:
Add the datamodel __match_args__ links in

dmoisset committed:
Refer to match wherever the C switch statement was compared with Python.

Brought up in https://bugs.python.org/msg387769

dmoisset committed:
Revert "object.__match_args__ -> class.__match_args__"

This reverts commit 6320747216b01cf78959d9209eb0ee1d89acc23d.

dmoisset committed:
Apply suggestions from code review

Some minor typos, and copy/paste fixes

dmoisset committed:
Refer to the PEP tutorial from the tutorial

dmoisset committed:
Fix documentation build errors

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

dmoisset committed:
Clean linter issues

Fidget-Spinner commented on Doc/reference/lexical_analysis.rst:
```diff
@@ -351,6 +351,25 @@ exactly as written here:
+
+The identifiers ``match``, ``case`` and ``_`` can syntactically act as keywords in some
+specific contexts related to the pattern matching statement, but this distinction is done
+at the parser level, not when tokenizing.
```
Extremely minor nit; feel free to ignore if you disagree. I think there should be at least a sentence describing what a _soft keyword_ actually is, similar to the keyword section above.

```suggestion
Some identifiers are only reserved under specific contexts. These are known as
*soft keywords*.  The identifiers ``match``, ``case`` and ``_`` can
syntactically act as keywords in contexts related to the pattern matching
statement, but this distinction is done at the parser level, not when
tokenizing.
```

dmoisset replied:
Nice phrasing, thanks!

Fidget-Spinner commented on Doc/reference/datamodel.rst:
```diff
@@ -2553,6 +2553,38 @@ For more information on context managers, see :ref:`typecontextmanager`.
+that ``case MyClass(x, y)`` is equivalent to ``case MyClass(left=x, center=y)``. Note
+that the number of arguments in the pattern must be smaller than or equal to the number
+of elements in *__match_args__*; if it is larger, the pattern match attempt will raise
+a *TypeError*.
```
```suggestion
a :exc:`TypeError`.
```

Fidget-Spinner commented on Doc/reference/datamodel.rst:
```diff
@@ -2553,6 +2553,38 @@ For more information on context managers, see :ref:`typecontextmanager`.
+   *__match_args__* as the keyword. The absence of this attribute is equivalent to
+   setting it to ``()``.
+
+For example if ``MyClass.__match_args__`` is ``("left", "center", "right")`` that means
```
Off topic: This example reads beautifully! It makes everything so easy to understand!

gvanrossum replied:
```suggestion
For example, if ``MyClass.__match_args__`` is ``("left", "center", "right")`` that means
```

Fidget-Spinner committed:
*TypeError* -> :exc:`TypeError` in Doc/reference/datamodel.rst

Fidget-Spinner committed:
Delete news file: shares same news as Brandt's PR

Fidget-Spinner committed:
update link in compound statements

dmoisset committed:
Better detail on what a soft keyword actually is

Co-authored-by: Ken Jin <28750310+Fidget-Spinner@users.noreply.github.com>

dmoisset committed:
Rework the sequence of steps in the overview

This avoid some possible confusion present in the previous language, where
it wasn't explicit that bindings happened before the guard was valuated

dmoisset committed:
Add some comments to the example

dmoisset committed:
Made the irrefutable definition slightly more stricter

dmoisset committed:
Move table rows as apragraphs into the description of each pattern

gvanrossum commented on Doc/faq/design.rst:
```diff
@@ -259,9 +259,8 @@ Why isn't there a switch or case statement in Python?
-consensus (yet) on whether and how to do range tests.  See :pep:`275` for
-complete details and the current status.
+For literal values, or constants within a namespace, you can also use a
+``match ... case`` statement.
```
I expect it would be better to rewrite or remove this section, but not now. IMO "match" *is* the missing switch/case statement.

willingc replied:
Or update the question to be: Is there a switch or case statement in Python?

gvanrossum commented on Doc/tutorial/controlflow.rst:
```diff
@@ -36,6 +36,9 @@ to avoid excessive indentation.  An  :keyword:`!if` ... :keyword:`!elif` ...
 :keyword:`!elif` ... sequence is a substitute for the ``switch`` or
 ``case`` statements found in other languages.
 
+If you're making comparison with constants, or checking for specific types or
```
1. "making comparison with" isn't a great phrase (I'd use the plural, "comparisons", or say "comparing to")
2. The match statement is really only useful if *the same* variable is used in each condition

```suggestion
If you're comparing the same value to several constants, or checking for specific types or
```

gvanrossum commented on Doc/tutorial/controlflow.rst:
```diff
@@ -246,6 +249,172 @@ at a more abstract level.  The :keyword:`!pass` is silently ignored::
    ...
 
+
+.. _tut-match:
```
Note that we now have two slightly diverging versions of the same introduction to match statements in the docs (three if you count Appendix A of PEP 636) -- the other is in what's new in 3.10. We don't have to fix this now but I think it would be good if the texts didn't arbitrarily differ from each other. Either the same text should be included in both places (possibly with a different intro and outro), or what's new should be much shorter and just point here. Again, doesn't need to be addressed today, but something to keep in mind. /CC @willingc

willingc replied:
It makes good sense to bring the two closer together. I would err on the side of clarity over length in the What's New as it's what most folks will read before or after the PEP tutorial.

gvanrossum reviewed (commented):
Thanks for this extensive work! I'll merge my two suggestions and then land it later tonight.

gvanrossum committed:
Apply my two specific suggestions from code review

willingc commented on Doc/reference/lexical_analysis.rst:
```diff
@@ -351,6 +351,27 @@ exactly as written here:
+Some identifiers are only reserved under specific contexts. These are known as
+*soft keywords*.  The identifiers ``match``, ``case`` and ``_`` can
+syntactically act as keywords in contexts related to the pattern matching
+statement, but this distinction is done at the parser level, not when
```
Good description.

willingc commented on Doc/reference/lexical_analysis.rst:
```diff
@@ -351,6 +351,27 @@ exactly as written here:
+statement, but this distinction is done at the parser level, not when
+tokenizing.
+
+This is done to allow their use while still
```
```suggestion
As soft keywords, their use with pattern matching is possible while still
```

willingc reviewed (commented):
Thanks @dmoisset. After the alpha goes out we can take a look at both this PR and the What's New again.

gvanrossum committed:
Final suggestion by Carol

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>

Fidget-Spinner committed:
delete redundant paragraph since table is gone
